src/models/StyledComponent.js

```diff
@@ -48,62 +51,111 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
       generatedClassName: '',
     }
 
+    constructor(props: Object, context: any) {
+      super(props, context)
+
+      let theme: any = null
+      let generatedClassName: string = ''
+
+      const styledContext = context[CHANNEL]
+      if (styledContext) {
+        const { getTheme } = styledContext
+        const contextTheme = getTheme()
+
+        // Props should take precedence over ThemeProvider, which should take precedence over
+        // defaultProps, but React automatically puts defaultProps on props.
+        const { defaultProps } = this.constructor
+        const isDefaultTheme = defaultProps && props.theme === defaultProps.theme
+        theme = props.theme && !isDefaultTheme ? props.theme : contextTheme
+        generatedClassName = this.generateAndInjectStyles(theme, props)
+      } else {
+        theme = props.theme || {}
+        generatedClassName = this.generateAndInjectStyles(
+          theme,
+          props,
+        )
+      }
+
+      this.state.theme = theme
+      this.state.generatedClassName = generatedClassName
+    }
+
     buildExecutionContext(theme: any, props: any) {
       const { attrs } = this.constructor
       const context = { ...props, theme }
       if (attrs === undefined) {
         return context
       }
 
-      this.attrs = Object.keys(attrs).reduce((acc, key) => {
+      const attrKeys = Object.keys(attrs)
+      for (let i = 0; i < attrKeys.length; i += 1) {
```

Comment:
This is an undesirable micro-optimisation :wink: While it will appear faster in micro-benchmarks `reduce` will perform almost identically :)

Also, this is mutating the `attrs` instance on the constructor, which will introduce a regression

A revert is preferable. Cheers :smile: